mantidproject/mantid
# Add tube calibration algorithms for Sans2d

**Description of work**

**Purpose of work**

The scripts used to calibrate the Sans2d instrument at ISIS were originally written by an instrument scientist. Whilst effective, they were complex and not well understood by other. Being so critical to the running of instrument, it became clear that they needed to be maintained within Mantid. This PR is for merging in a first version of those scripts, refactored into Mantid algorithms with changes to try and improve maintainability, along with tests and documentation.

There are a large number of commits in this PR. For the record, the first commit preserves the first version of the algorithms that we were able to get working, but that were still largely created by copying and pasting the original instrument scientist scripts. We wanted to keep these in case we ever want to reinstate anything that was in the original script, but that we don't currently think we need (for example, the option to use a Gaussian fit for peak fitting).

**Report to:** Leide at ISIS

**Summary of work**

This PR includes the following:

- `SANSTubeCalibration` - the main algorithm used to perform a calibration of either the front or rear Sans2d detector.
- `SANSTubeMerge` - the algorithm used to merge the separate front and rear detectors into a single calibration file.
- `FlatTopPeak` - a new fitting function that is used to fit the peak positions in each detector tube.
- System and unit tests, as relevant, for the new algorithms and function.
- Documentation for `SANSTubeCalibration` and `SANSTubeMerge`. The data required to run these algorithms takes a long time to load, so I have provided a Usage section in the docs for `SANSTubeCalibration` that is slightly different to other algorithms as I don't think we want the doctests running on this (unless I'm misunderstanding how these tests work). I haven't added a usage section for the `SANSTubeMerge` docs as the same loading time issues apply and the algorithm is straight-forward to use. I haven't included documentation for `FlatTopPeak` as I don't understand this sufficiently well to be able to explain it - we will need to speak to our scientists/any developers on the team with a science background to add this in a future PR.

This PR should be considered a first step but with further improvements to come - the aim is to have something in the code that we feel we can maintain, but I still consider it to be far from perfect. See the issues linked from #27386 that describe the outstanding tasks still on our list. It has already taken a significant amount of time to get to this stage, and unfortunately we are limited in the amount of time we have available in LSS to make quick progress on this, so one of the issues I've opened to capture outstanding tasks is focussed on maintenance - see #35982. There is benefit to having the first version of these changes in Mantid, so while I'm expecting there to be several suggestions for improving maintainability and validation as part of this review, I would be grateful if we could add anything that's not critical to the opened maintenance issue to be addressed as part of ongoing work, rather than hold up this PR.


**To test:**


The algorithms have been tested extensively by our instrument scientist contact in SANS and are actually being used to perform their calibration at the moment (by adding the scripts to Mantid via the Python Script Directories in the Manage User Directories dialog). I've tried to add relevant system tests and the load time for the data is very slow when run on a local machine, so I think simply checking that the algorithms can be looked up from the Algorithms pane of Mantid Workbench would be sufficient for this PR.

If keen to run something, though, then the usage example provided in the `SANSTubeCalibration` documentation would be a good option. The scientists are typically running this via Mantid's algorithm execution dialog at the moment, rather than from a script, so reading through the documentation and then running the usage example via that dialog would be a good test. Otherwise reading through the documentation to check that it makes some sense, and ensuring test coverage seems sufficient would be good (note I haven't been able to add as many unit tests as would be ideal because of the time taken to load Sans2d data).

This packaging build is to confirm that the doctests are all passing OK: https://builds.mantidproject.org/job/build_packages_from_branch/552/

Fixes #27386. 






---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
